File: scripts/python/automata.py
#automata.py
"""
A script containing Grid and Cell clases for the automata challenge
"""

class Cell:
	
	def __init__(self, x, y, state):
		self.x = x
		self.y = y
		self.state = state

class GridState:

	def __init__(self, size):
		#self.cells = [Cell(0,0,False) for x in range(size*size)]
		self.cells = []
		self.size = size
		self.length = size*size

	def getCellState(self, x, y):
		return self.cells[x + y*self.size].state

	def setCellState(self, x, y, state):
		self.cells[x + y*self.size].state = state

	def nextGen(self):
		nCells = []
		for cell in self.cells:
			live = self.live_neighbors(cell.x,cell.y)
			if(live == 1 or live == 2):
				nCells.append(Cell(cell.x, cell.y, True))
			else:
				nCells.append(Cell(cell.x, cell.y, False))
		self.cells = nCells

	def live_neighbors(self, i, j):
		live = 0
		nx = [i-1,i-1,i-1,i,i,i+1,i+1,i+1]
		ny = [j-1,j,j+1,j-1,j+1,j-1,j,j+1]	
		for x, y in zip(nx,ny):
			if(x>=0 and x<self.size and y>=0 and y<self.size):
				if(self.cells[y + x*self.size].state):
					live += 1
		return live
	
	def state2Int(self):
		strState = ''
		for cell in self.cells:
			strState = strState + str(int(cell.state))

		return int(strState,2)

	def stateFromInt(self, num):
		strnum = "{0:b}".format(num).zfill(self.length)

		index = 0
		i = 0
		j = 0
		while(index < self.length):
			self.cells.append(Cell(i,j,bool(int(strnum[index]))))
			"""
			self.cells[index].state = bool(int(strnum[index]))
			self.cells[index].x = i
			self.cells[index].y = j
			"""
			
			j = j + 1
		
			if(j > self.size-1):
				j = 0
				i += 1
			index += 1

	def isExtinct(self):
		return self.state2Int() == 0

	def printState(self):
		print("----")
		for i in range(self.size):
			for j in range(self.size):
				print(int(self.cells[j+i*self.size].state), end='')
			print("\n")
		print("----")

File: src/automata.cpp
#include "automata.h"

gridState::gridState(int size){
	this->size = size;
	this->length = size*size;
}

void gridState::createFromInt(int num){
	std::vector<int> state;
	
	while(num > 0){
		state.insert(state.begin(), (int) num%2);
		num = (int) floor(num/2);
	}

	if((int) state.size() != this->length){
		int i = this->length - (int) state.size();
		while(i > 0){
			state.insert(state.begin(), 0);
			i--;
		}
	}
	this->cells = state;
}

int gridState::getSize(void){
	return this->size;
}

int gridState::getLen(void){
	return this->length;
}

void gridState::printState(void){
	if(sizeof(this->cells) != 0){
		std::cout<<"State: "<<std::endl;
		for(int j=0; j < this->size; j++){
			for(int i=0; i < this->size; i++){
				std::cout<<this->cells[i+(j*this->size)]<<" ";
			}
			std::cout<<std::endl;
		}
	}else{
		std::cout<<"State is empty!"<<std::endl;
	}
}

int gridState::state2Int(void){
	int num = 0;
	int weight = this->length - 1;
	for(int i=0; i < this->length; i++){
		num = num + (pow(2,weight)*cells[i]);
		weight--;
	}
	return num;
}

int gridState::live_neighbors(int i, int j){
	int live = 0;
	int const n = 8;
	std::array<int, n> x; 
	std::array<int, n> y;
	x[0] = i-1;
	x[1] = i-1;
	x[2] = i-1;	
	x[3] = i;
	x[4] = i;
	x[5] = i+1;
	x[6] = i+1;
	x[7] = i+1;
	
	y[0] = j-1;
	y[1] = j;
	y[2] = j+1;
	y[3] = j-1;
	y[4] = j+1;
	y[5] = j-1;
	y[6] = j;
	y[7] = j+1;
	
	for(int t = 0; t < n; t++){
		if(x[t]>=0 && x[t] < this->size && y[t]>=0 && y[t] < this->size){
			if(this->cells[y[t]+(x[t]*this->size)] == 1){
				live++;
			}
		}
	}
	return live;
}

void gridState::nextGen(void){
	std::vector<int> nState;
	
	for(int i=0; i<this->size; i++){
		for(int j=0; j<this->size; j++){
			int nb_sum = this->live_neighbors(i,j);
			if(nb_sum == 1 || nb_sum == 2){
				nState.push_back(1);
			}else{
				nState.push_back(0);
			}
		}
	}
	this->cells = nState;
}
